WIP / ENH: spatial.transform.Spline [GSOC 2018]

adbugger opened the pull request:
@nmayorov , preprocessing during initialization of class has been implemented.

rgommers commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1590,3 +1590,202 @@ def __call__(self, times):
+        w_int, diff = _intermediate_step(np.zeros(len(rotations) - 2, 3),
+                                         rotvecs, dt_inv)
+
+        while diff >= 1e-6:
```
Should this 1e-6 magic number be configurable? and why is 1e-6 the right number?

adbugger replied:
This `1e-6` was mentioned in the paper which Nikolay provided. I'm not sure it suits all applications. Should I add a tolerance argument to the function and explain the convergence steps in the docstring?

adbugger commented:
Fixed a few bugs regarding intermediate angular velocity calculation and intermediate error calculation.

Added a `tol` variable to compute the error tolerance in angular velocity. Set default as `1e-12`. Paper mentions `1e-6` but only as an estimate of number of iterations. In practice, the number of steps taken to reach convergence depends on the 'non-linearity' or the 'jaggedness' of the path which depends on the orientations, the times and the specified initial and final angular velocities.

On a jagged path generated by 10 random rotations, `~100` iterations yield an additional accuracy of 6 orders of magnitude. Have chosen default value of `tol` as `1e-12` to converge at `~200` iterations. Let me know if that seems acceptable.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1639,3 +1639,263 @@ def __call__(self, times):
+        self.omega[-1] = omega_f
+        self.omega[1:-1] = w_int
+
+        # Formula for theta(t) from eqn 16, coefficients from page 5 of ref
```
Let's use [http://scipy.github.io/devdocs/generated/scipy.interpolate.PPoly.html#scipy.interpolate.PPoly](PPoly) to store coefficients and do polynomial evaluations.

adbugger replied:
This will require some re-formulation of the polynomial in the paper, I'll do that.

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -889,3 +889,80 @@ def test_slerp_call_time_out_of_range():
         s([1, 2, 6])
+
+
+def test_spline_trivial():
```
Add non-trivial tests.

Try to find a way to somehow check the continuity of the solution.

adbugger replied:
How about calculating angular velocity (and angular acceleration?) close to the break points using a finite difference approximation?

adbugger replied:
So I implemented this method and it looks like there are some bugs which still need to be ironed out. The angular velocities at `key_times + delta_t` agree to within `1e-3`, but the velocities at `key_times - delta_t` do not. I'll keep looking into it.

nmayorov replied:
It might be not exactly a bug. What I suggest to do first is to implement angular velocity and acceleration computations (suggested in my email) and plot them as the function of time. It will be clear whether the behavior is correct. If it is correct, you will find a way to check it formally in tests.

adbugger commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1639,3 +1639,263 @@ def __call__(self, times):
                 Rotation.from_rotvec(self.rotvecs[ind] * alpha[:, None]))
+
+
+def _Rfunc(rotvec, omega):
```
The thing is that the functions are quite simply replacements for long terms in the equations. I'm not sure if their purpose can be put in a succinct name. I'll go over the paper and try again.

adbugger commented:
So I've used `scipy.interpolate.PPoly` for various interpolations. One conceptual error I was making earlier was confusing `theta_dot` from the paper with the angular velocity vector. Further, calculating the angular velocity vector from quaternion rotations is not as straightforward as using the earlier finite difference approximation. For now I have just stored the coefficients for `theta_dot` and `theta_dot2`. Computing the angular velocity and angular acceleration from them requires some more work. I'm on it right now.

adbugger commented:
Figured out the bugs. Added a function for returning angular velocity and renamed to QSpline. Angular velocity continuity is now satisfied for smooth changes in orientation; for random rotations continuity at the right ends of the intervals is still sometimes an issue, however looking at the graphs in the paper it appears that problem cannot be completely removed.

@nmayorov the `from scipy.interpolate.PPoly` fails on CIs. What could be the issue there? The tests pass locally.

ewmoore commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1639,3 +1640,308 @@ def __call__(self, times):
+    c2 = np.empty_like(c1)
+    c2[~small_theta] = (1 - np.cos(large_angle)) / large_angle
+    c2[small_theta] = small_angle / 2 - (small_angle ** 3) / 24
+
```
The need to do this small vs. large angle calculation for `c1` and `c2` can probably be replaced by `special.sinc` (or `np.sinc`) and `special.cosm1`.

nmayorov replied:
@ewmoore nice suggestion!

Certainly can be used in a lot of places (including some previous code @adbugger wrote). In this particular case we should represent 1 - cos(x) as 2 sin^2 (x/2) and then sinc emerges.

One thing which is slightly annoying is that scipy/numpy defines sinc as sin(pi x) / (pi x), I would much prefer to have a clean definition sin x / x. I guess we may write a simple wrapper to alleviate that.

@adbugger investigate that idea. I think we can remove all small/large angle branches using sinc, here and in `Rotation` class as well.

adbugger replied:
Will implement this in a separate PR.

rgommers commented:
> Any ideas on how to work around it?

Removing `from . import transform` in `spatial/__init__.py` may be enough. There's anyway a case that can be made for that.

Another option is to only import `interpolate` within `QSpline`. In general it's good if scipy submodules are not all dependent on each other; and here the use of `interpolate` within `spatial` is (a) new and (b) very limited.

rgommers commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1639,3 +1640,308 @@ def __call__(self, times):
+    return w, diff
+
+
+class QSpline(object):
```
How about calling this `QuaternionSpline`? Or `QuaternionInterpolator`? Or yet something else - just want to avoid giving the impression that this is a spline thingy that can be reused (it's very similar to B-spline now).

adbugger replied:
How about renaming both functions to `RotationSlerp` and `RotationSpline`? In my mind this makes is explicit that these classes are for use only with `Rotation`s, as opposed to general interpolators for quaternions.

nmayorov replied:
`RotationSlerp` and `RotationSpline` sounds logical to me as well.

tylerjereddy commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -889,3 +889,117 @@ def test_slerp_call_time_out_of_range():
+    with pytest.raises(ValueError, match="times must be within the range"):
+        s([0, 1, 2])
+    with pytest.raises(ValueError, match="times must be within the range"):
+        s([1, 2, 6])
```
Could reduce code duplication here with `@pytest.mark.parametrize`

tylerjereddy commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -889,3 +889,117 @@ def test_slerp_call_time_out_of_range():
+
+    vel = spline.omega[1]
+    assert_allclose(vel, omega_i)
+    assert_allclose(vel, omega_f)
```
`omega_i` and `omega_f` are identical list objects in this unit test; why the assertion for `vel` compared to both as apposed to a single reference value? Are you trying to ensure that `QSpline` doesn't modify the omega argument objects passed in?

adbugger replied:
You're right. Only one check is required here.

adbugger commented:
@nmayorov , @rgommers is this mergable? I'll investigate the idea of using `np.special.sinc` in a separate PR.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1639,3 +1639,370 @@ def __call__(self, times):
+
+        return initial_rot * interpolating_quat
+
+    def angular_velocity(self, times):
```
I think it should go to `__call__` with some argument like `order`.

nmayorov commented:
@adbugger I'm not 100% satisfied with the code. I will try to review thoroughly on weekends. Meanwhile I think you should also implement computation of the angular acceleration (since we guarantee its continuity --- it will be interesting to at least see that). 

I suggest the interface `__call__(times, order=0)`

adbugger commented:
Alright, I'll get on it. It's just that on 6 Aug the official coding period ends and the evaluation period starts (6 - 14 Aug). So, depending on the changes required, a review over the weekend _might_ be cutting it a little too close.

I'll implement the `__call__` interface and the acceleration functionality.

nmayorov commented:
Experimented a bit with the `RotationSpline`: https://gist.github.com/nmayorov/871380118428ae9e41bf6e4dd36cd35a

Computation of angular velocity and acceleration seems to be incorrect. There should not be any breaks in lines.

nmayorov commented:
Closed in favor of #9024.
